Add _replica and _replica_first as search preference.

dakrone opened the pull request:
Just like specifying `?preference=_primary`, this adds the ability to
specify `?preference=_replica` or `?preference=_replica_first` on
requests that support it.

Resolves #12222

kimchy commented on core/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java:
```diff
@@ -312,6 +316,32 @@ public ShardIterator primaryFirstActiveInitializingShardsIt() {
     }
 
+    public ShardIterator replicaActiveInitializingShardIt() {
+        if (!replicas.isEmpty() && !replicas.get(0).active() && !replicas.get(0).initializing()) {
```
I wouldn't rely on the order the replicas exist in for this behavior, note it works in the primary case cause there can be only one :)

kimchy commented on core/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java:
```diff
@@ -312,6 +316,32 @@ public ShardIterator primaryFirstActiveInitializingShardsIt() {
+                ordered.set(0, shardRouting);
+            }
+        }
+        // no need to worry about replica first here..., its temporal
```
we know there is only one primary, so we can optimize this case and push it to the end once we are done?

kimchy commented:
left a couple of comments, can we also add unit tests to `RoutingIteratorTests`? I think in that case, we don't really need to have an integration test

dakrone commented:
@kimchy thanks for taking a look! I think I misunderstood what the iterators were doing previously, but I think I have fixed it now. (and added a unit test to make sure)

kimchy commented on core/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java:
```diff
@@ -312,6 +312,55 @@ public ShardIterator primaryFirstActiveInitializingShardsIt() {
+    public ShardIterator replicaActiveInitializingShardIt() {
+        // If the primaries are unassigned, return an empty list (there aren't
+        // any replicas to query anyway)
+        if (!primaryAsList.isEmpty() && !primaryAsList.get(0).active() && !primaryAsList.get(0).initializing()) {
```
can me extract this into a method, it is used in 3 places

dakrone replied:
Good idea, I'll do that.

kimchy commented on core/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java:
```diff
@@ -312,6 +312,55 @@ public ShardIterator primaryFirstActiveInitializingShardsIt() {
+
+        ArrayList<ShardRouting> activeReplicas = new ArrayList<>();
+        ArrayList<ShardRouting> initializingReplicas = new ArrayList<>();
+        ArrayList<ShardRouting> ordered = new ArrayList<>(activeReplicas.size());
```
optimization nit, but maybe we can just have one list, and reorder to push active one to the start, we do something similar in primaryFirst. This will mean we don't have to create 3 lists, just one

dakrone replied:
Will do.

kimchy commented on core/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java:
```diff
@@ -312,6 +312,55 @@ public ShardIterator primaryFirstActiveInitializingShardsIt() {
+        }
+
+        // Add the rest of the primary shards, shuffled
+        ordered.addAll(shuffler.shuffle(primaryAsList));
```
no need to shuffle here, since it is always a single element list

dakrone replied:
ahh good point.

kimchy commented:
left few minor comments, LGTM
